assets/src/dashboard/components/storyAnimation/provider.js

```diff
@@ -0,0 +1,160 @@
+
+const WAAPIAnimationMachine = {
+  idle: {
+    complete: 'complete',
+  },
+  complete: {
+    reset: 'idle',
+  },
+};
+
+const WAAPIAnimationStateReducer = (state, action) => {
+  return WAAPIAnimationMachine[state][action] || state;
+};
+
+const createOnFinishPromise = (animation) => {
+  return new Promise((resolve) => {
+    animation.onfinish = resolve;
+  });
+};
+
+function Provider(props) {
+  const animationGeneratorMap = useMemo(() => {
+    return (props.animations || []).reduce((map, animation) => {
+      const { target, type, ...args } = animation;
+
+      (Array.isArray(target) ? target : [target]).forEach((t) => {
+        const animationGenerators = map.get(t) || [];
+        map.set(t, [...animationGenerators, (fn) => fn(type, args)]);
+      });
+
+      return map;
+    }, new Map());
+  }, [props.animations]);
+
+  const getAnimationGenerators = useCallback(
+    (target) => {
+      return animationGeneratorMap.get(target) || [];
+    },
+    [animationGeneratorMap]
+  );
+
+  /**
+   * WAAPI interface
+   */
+  const onWAAPIFinishRef = useRef(props.onWAAPIFinish);
+  const [WAAPIAnimationState, dispatchWAAPIAnimationState] = useReducer(
+    WAAPIAnimationStateReducer,
+    'idle'
+  );
+  const WAAPIAnimationMap = useMemo(() => new Map(), []);
+  const [WAAPIAnimations, setWAAPIAnimations] = useState([]);
+
+  const hoistWAAPIAnimation = useCallback(
+    (WAPPIAnimation) => {
+      const symbol = Symbol();
+      WAAPIAnimationMap.set(symbol, WAPPIAnimation);
+      setWAAPIAnimations(Array.from(WAAPIAnimationMap.values()));
+      return () => {
+        WAPPIAnimation?.cancel();
+        WAAPIAnimationMap.delete(symbol);
+        setWAAPIAnimations(Array.from(WAAPIAnimationMap.values()));
+      };
+    },
+    [WAAPIAnimationMap]
+  );
+
+  const playWAAPIAnimations = useCallback(() => {
+    WAAPIAnimations.forEach((animation) => animation?.play());
+  }, [WAAPIAnimations]);
+
+  /**
+   * Browser support for `animation.finished` is no good.
+   * https://developer.mozilla.org/en-US/docs/Web/API/Animation/finished
+   *
+   * So we're mimicking the spec by creating a new promise everytime all
+   * animations complete.
+   */
+  useEffect(() => {
+    // WAAPIAnimations.forEach((animation) => {
+    //   // createOnFinishPromise
+    //   animation.onfinish = onWAAPIFinishRef.current;
+    // });
+    let cancel = () => {};
+    if ('idle' === WAAPIAnimationState && WAAPIAnimations.length) {
+      new Promise((resolve, reject) => {
+        cancel = reject;
+        Promise.all(WAAPIAnimations.map(createOnFinishPromise)).then(() => {
+          cancel = () => {};
+          resolve();
+        });
+      })
+        .then(() => dispatchWAAPIAnimationState('complete'))
+        /* needed if promise gets canceled to swallow the error */
+        .catch(() => {});
+    }
+    return cancel;
+  }, [WAAPIAnimations, WAAPIAnimationState]);
+
+  onWAAPIFinishRef.current = props.onWAAPIFinish;
```

Comment:
Yea so the line `const onWAAPIFinishRef = useRef(props.onWAAPIFinish);` only sets the ref on mount. 

rn I have `onWAAPIFinishRef.current = props.onWAAPIFinish;` because I want it to be up to date on every render, but I don't want it to cause rerenders in the hook below it.

Reply:
Oooh, interesting 🤔 That makes sense.